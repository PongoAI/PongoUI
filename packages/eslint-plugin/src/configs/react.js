/** @type {import("eslint").Linter.Config} */
const config = {
  extends: ['airbnb-typescript', 'prettier'],
  parser: '@typescript-eslint/parser',
  parserOptions: {
    project: 'tsconfig.json',
  },
  plugins: ['deprecation', 'import', '@typescript-eslint', 'jest', 'jsx-a11y', 'react', 'react-hooks'],
  settings: {
    'import/parsers': {
      '@typescript-eslint/parser': ['.ts', '.tsx'],
    },
    'import/resolver': {
      typescript: {
        alwaysTryTypes: true,
      },
    },
    'no-unused-expressions': 'off',
  },
  env: {
    browser: true,
    es6: true,
    node: true,
  },

  ignorePatterns: ['coverage', 'dist', 'dist-storybook', 'lib', 'lib-commonjs', '**/__snapshots__', 'node_modules'],
  rules: {
    'no-unused-vars': 'error',
    'no-use-before-define': 'off',
    '@typescript-eslint/no-use-before-define': ['error'],
    'import/no-extraneous-dependencies': ['error', { devDependencies: true }],
    'import/extensions': [
      'error',
      'ignorePackages',
      {
        js: 'never',
        jsx: 'never',
        ts: 'never',
        tsx: 'never',
      },
    ],
  },
};

module.exports = config;
