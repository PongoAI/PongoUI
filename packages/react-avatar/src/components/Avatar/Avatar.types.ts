import type { ComponentProps, ComponentState, Slot } from '@fluentui/react-utilities';
import type { ColorTokens } from '@cebus/react-theme';

export type AvatarSlots = {
  /**
   * The root of the Avatar.
   */
  root: Slot<'span'>;

  /**
   * The initials to display within the avatar or use as a fallback. Initials come from the name prop.
   */
  initials?: Slot<'span'>;

  /**
   * The image to display for the Avatar.
   */
  image?: Slot<'img'>;
};

export type AvatarCommons = {
  /**
   * The name of the user to display or use as a fallback when using an image.
   *
   * @defaultValue "?"
   */
  name?: string;

  /**
   * The shape of the Avatar.
   *
   * @defaultValue circular
   */
  shape?: 'circular' | 'square';

  /**
   * The size of the Avatar.
   *
   * @defaultValue medium
   */
  size?: 'small' | 'medium' | 'large' | 'extra-large';

  /**
   * The color to render the background of the icon. By default the color will be chosen based on the name prop.
   */
  color?: keyof ColorTokens;
};

export type AvatarProps = ComponentProps<AvatarSlots> & AvatarCommons;

export type AvatarState = ComponentState<AvatarSlots> & AvatarCommons;
