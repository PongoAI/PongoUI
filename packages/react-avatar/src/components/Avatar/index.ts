export * from './Avatar';
export * from './useAvatarStyles';
