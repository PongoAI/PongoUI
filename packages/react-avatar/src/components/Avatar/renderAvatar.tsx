import * as React from 'react';
import { getSlots } from '@fluentui/react-utilities';
import type { AvatarState, AvatarSlots } from './Avatar.types';

export const renderAvatar = (state: AvatarState) => {
  const { slots, slotProps } = getSlots<AvatarSlots>(state);

  return (
    <slots.root {...slotProps.root}>
      {slots.initials && <slots.initials {...slotProps.initials} />}
      {slots.image && <slots.image {...slotProps.image} />}
    </slots.root>
  );
};
