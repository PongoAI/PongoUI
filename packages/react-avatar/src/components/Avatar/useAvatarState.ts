import * as React from 'react';
import type { AvatarState } from './Avatar.types';
import { useMergedEventCallbacks } from '@fluentui/react-utilities';
import { getInitials } from '@fluentui/react-avatar';
import { stringToColor } from '../../utils/index';
import { useFluent } from '@fluentui/react-shared-contexts';
import { tokens } from '@cebus/react-theme';
import { calculateTransparentColor } from '@cebus/react-theme-generator';

export const useAvatarState = (state: AvatarState): AvatarState => {
  const [imageHidden, setImageHidden] = React.useState<true | undefined>(undefined);

  const { name, color } = state;
  const { dir } = useFluent();

  const userInitials = name ? getInitials(name, dir === 'rtl') : '?';
  // Darken the generated color to help prevent contrast ratio issues with the white text.
  const avatarGeneratedColor = name ? calculateTransparentColor('#000000', stringToColor(name), 0.3) : tokens.inherit;

  const imageOnErrorCallback = useMergedEventCallbacks(state.image?.onError, () => setImageHidden(true));

  const imageOnLoadCallback = useMergedEventCallbacks(state.image?.onLoad, () => setImageHidden(undefined));

  const rootStyles = {
    background: color ? tokens[color] : avatarGeneratedColor,
  };

  // Root Props
  state.root.style = { ...rootStyles, ...state.root.style };
  if (!state.root['aria-label'] && !state.root['aria-labelledby'] && name) {
    state.root['aria-label'] = name;
  }

  // Initials Props
  if (state.initials) {
    state.initials.children = userInitials;
  }

  // Image Props
  if (state.image) {
    state.image.hidden = imageHidden;
    state.image.onError = imageOnErrorCallback;
    state.image.onLoad = imageOnLoadCallback;
  }

  return state;
};
