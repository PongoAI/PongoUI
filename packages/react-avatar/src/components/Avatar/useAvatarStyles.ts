import { makeStyles, mergeClasses, shorthands } from '@griffel/react';
import type { AvatarState } from './Avatar.types';
import { tokens } from '@pongo-ui/react-theme';

const useStyles = makeStyles({
  root: {
    position: 'relative',
    fontFamily: tokens.baseFont,
    fontWeight: tokens.medium,
    verticalAlign: 'middle',
    color: 'white',
    display: 'flex',
    userSelect: 'none',
    flexShrink: 0,
    flexGrow: 0,
    alignItems: 'center',
    justifyContent: 'center',
    boxSizing: 'border-box',
    ...shorthands.padding('0px'),
    ...shorthands.margin('0px'),
  },

  image: {
    position: 'absolute',
    top: 0,
    left: 0,
    width: '100%',
    height: '100%',
    objectFit: 'cover',
    verticalAlign: 'top',
  },

  circular: { ...shorthands.borderRadius(tokens.circular) },

  square: { ...shorthands.borderRadius(tokens.square) },

  initials: {
    ...shorthands.padding('0px'),
    ...shorthands.margin('0px'),
  },

  small: {
    width: '28px',
    height: '28px',
    fontSize: tokens.fontSize300,
  },

  medium: {
    width: '35px',
    height: '35px',
    fontSize: tokens.fontSize400,
  },

  large: {
    width: '50px',
    height: '50px',
    fontSize: tokens.fontSize600,
  },

  'extra-large': {
    width: '70px',
    height: '70px',
    fontSize: tokens.fontSize800,
  },
});

export const useAvatarStyles = (state: AvatarState) => {
  const styles = useStyles();

  state.root.className = mergeClasses(styles.root, styles[state.shape!], styles[state.size!], state.root.className);

  if (state.initials) {
    state.initials.className = mergeClasses(styles.initials, state.initials.className);
  }

  if (state.image) {
    state.image.className = mergeClasses(styles.image, styles[state.shape!], state.image.className);
  }

  return state;
};
