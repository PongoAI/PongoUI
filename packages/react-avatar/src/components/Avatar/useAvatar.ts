import * as React from 'react';
import { getNativeElementProps, resolveShorthand, useId } from '@fluentui/react-utilities';
import { useAvatarState } from './useAvatarState';
import type { AvatarProps, AvatarState } from './Avatar.types';

export const useAvatar = (props: AvatarProps, ref: React.Ref<HTMLElement>): AvatarState => {
  const { name, color, shape = 'circular', size = 'medium', initials, image } = props;

  const avatarId = useId('avatar-');

  const state: AvatarState = {
    name,
    shape,
    size,
    color,
    components: {
      root: 'span',
      initials: 'span',
      image: 'img',
    },
    root: getNativeElementProps('span', {
      role: 'img',
      id: avatarId,
      ...props,
      ref,
    }),
    image: resolveShorthand(image, {
      required: false,
      defaultProps: {
        alt: '',
        role: 'presentation',
        'aria-hidden': true,
      },
    }),
    initials: resolveShorthand(initials, {
      required: true,
      defaultProps: {
        'aria-hidden': true,
      },
    }),
  };

  useAvatarState(state);

  return state;
};
