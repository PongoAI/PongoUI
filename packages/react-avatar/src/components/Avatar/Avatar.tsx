import * as React from 'react';
import { useAvatar } from './useAvatar';
import { renderAvatar } from './renderAvatar';
import { useAvatarStyles } from './useAvatarStyles';
import type { AvatarProps } from './Avatar.types';
import type { ForwardRefComponent } from '@fluentui/react-utilities';

/**
 * The Avatar provides a recognizable icon for a user.
 */
export const Avatar: ForwardRefComponent<AvatarProps> = React.forwardRef((props, ref) => {
  const state = useAvatar(props, ref);

  useAvatarStyles(state);

  return renderAvatar(state);
});
Avatar.displayName = 'Avatar';
