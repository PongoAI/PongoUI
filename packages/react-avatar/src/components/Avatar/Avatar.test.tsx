import * as React from 'react';
import { Avatar } from './Avatar';
import { isConformant } from '../../common/isConformant';
import { render } from '@testing-library/react';

describe('Avatar', () => {
  isConformant({
    Component: Avatar,
    displayName: 'Avatar',
  });

  describe('Snapshot tests', () => {
    it('renders a basic Avatar', () => {
      const result = render(<Avatar name="test" />);

      expect(result.baseElement).toMatchSnapshot();
    });

    it('renders an Avatar with an image', () => {
      const result = render(
        <Avatar
          name="test"
          image={{ src: 'https://th.bing.com/th/id/OIP.R3IAG0Z7SAzTV56Xuvs1LwHaFj?pid=ImgDet&rs=1' }}
        />,
      );

      expect(result.baseElement).toMatchSnapshot();
    });
  });
});
