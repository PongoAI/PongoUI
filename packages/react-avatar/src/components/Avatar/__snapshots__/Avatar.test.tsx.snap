// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Avatar Snapshot tests renders a basic Avatar 1`] = `
<body>
  <div>
    <span
      aria-label="test"
      class="___dd042x0 f10pi13n fk6fouc ffu84w5 fmrv4ls fmysjtr f22iagw f1xqy1su fi64zpg f1jhi6b8 f122n59 f4d9j23 f1ewtqcl f1nbblvp fifp7yv f1ov4xf1 f1asdtw4 f1lmfglv fibjyge f1abmfm4 f9yszdx f1i2uu5k f1ys4a3y f1gczwgj fadrijh f1bc5yav fcv2ht4 f1mg03ek"
      id="avatar-7"
      name="test"
      role="img"
      style="background: rgb(102, 48, 38);"
    >
      <span
        aria-hidden="true"
        class="___bbcx9p0 f1nbblvp fifp7yv f1ov4xf1 f1asdtw4 f1lmfglv fibjyge f1abmfm4 f9yszdx"
      >
        T
      </span>
    </span>
  </div>
</body>
`;

exports[`Avatar Snapshot tests renders an Avatar with an image 1`] = `
<body>
  <div>
    <span
      aria-label="test"
      class="___dd042x0 f10pi13n fk6fouc ffu84w5 fmrv4ls fmysjtr f22iagw f1xqy1su fi64zpg f1jhi6b8 f122n59 f4d9j23 f1ewtqcl f1nbblvp fifp7yv f1ov4xf1 f1asdtw4 f1lmfglv fibjyge f1abmfm4 f9yszdx f1i2uu5k f1ys4a3y f1gczwgj fadrijh f1bc5yav fcv2ht4 f1mg03ek"
      id="avatar-8"
      name="test"
      role="img"
      style="background: rgb(102, 48, 38);"
    >
      <span
        aria-hidden="true"
        class="___bbcx9p0 f1nbblvp fifp7yv f1ov4xf1 f1asdtw4 f1lmfglv fibjyge f1abmfm4 f9yszdx"
      >
        T
      </span>
      <img
        alt=""
        aria-hidden="true"
        class="___1xbm6i0 f1euv43f f15twtuk f1vgc2s3 fly5x3f f1l02sjl f1ps3kmd f12kltsn f1i2uu5k f1ys4a3y f1gczwgj fadrijh"
        role="presentation"
        src="https://th.bing.com/th/id/OIP.R3IAG0Z7SAzTV56Xuvs1LwHaFj?pid=ImgDet&rs=1"
      />
    </span>
  </div>
</body>
`;
